### PR TITLE
Allow RL to run inference-only via skip-train

### PR DESCRIPTION
**Pull request by tdene**
# What does this PR do ?

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact the @mcore-oncall.

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment the [@mcore-oncall](https://github.com/orgs/NVIDIA/teams/mcore-oncall) to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### CODE REVIEW

File: megatron/training/arguments.py [hunk 1]
```diff
@@ -2212,6 +2224,9 @@ def _add_rl_args(parser):
                        help="Default top-k for model inference.")
     group.add_argument('--rl-offload-optimizer-during-inference', action='store_true',
                        help='Offload optimizer state to CPU during inference/rollout to save GPU memory')
+    group.add_argument('--rl-skip-optimizer', action=argparse.BooleanOptionalAction, default=False,
```
Comment (jaredcasper) [inline megatron/training/arguments.py]: When would I want to create an optimizer when in inference only mode?

Comment (tdene) [reply megatron/training/arguments.py]: * Gauge memory pressure whilst finetuning inference params
 * Test optimizer offload / onload
 * Simulate the time pressure of all the onload / offloads together (model swap, optimizer, inference)

None of these are strong reasons, but it's easy enough to offer the option. And it was requested by a reviewer.

File: megatron/training/training.py [hunk 2]
```diff
@@ -2880,20 +2902,31 @@ def trace_handler(p):
...
-            should_exit,
-            exit_code,
-            grad_norm,
-            num_zeros_in_grad,
-            max_attention_logit,
-        ) = train_step(
-            forward_step_func, train_data_iterator, model, optimizer, opt_param_scheduler, config, forward_backward_func, iteration=iteration
-        )
-        ft_integration.on_training_step_end()
+        if args.skip_train:
```
[the remark this answers is not in the document]
Comment (tdene) [reply megatron/training/training.py:2905]: I thank you for your detailed feedback, Claude, but please consider that this is the intended outcome of this code:

> if a future caller or configuration ends up invoking train() with skip_train=True but without perform_rl_step, this would silently skip all training steps every iteration.

File: megatron/training/training.py [hunk 3]
```diff
@@ -3479,14 +3515,14 @@ def build_train_valid_test_data_loaders(build_train_valid_test_datasets_provider
...
+        if args.perform_rl_step:
             # we don't need to build any dataloaders for RL training
             train_dataloader = None
             valid_dataloaders = None
             test_dataloader = None
-            do_train = args.train_iters > 0
+            do_train = (args.train_iters or 0) > 0
             do_valid = (args.full_validation or args.eval_iters > 0)
-            do_test = (args.full_validation or args.eval_iters > 0)
+            do_test = False
```
[the remark this answers is not in the document]
Comment (tdene) [reply megatron/training/training.py]: You're right Claude; there's no point in making this diff in this PR. Addressed.
